internal/core/infrastructure/feeder/kraken/service.go

```diff
@@ -0,0 +1,198 @@
+package krakenfeeder
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/gorilla/websocket"
+	"github.com/shopspring/decimal"
+	log "github.com/sirupsen/logrus"
+	"github.com/tdex-network/tdex-feeder/internal/core/ports"
+)
+
+const (
+	// KrakenWebSocketURL is the base url to open a connection with kraken.
+	// This can be tweaked if in the future it might change, even if unlikely.
+	KrakenWebSocketURL = "ws.kraken.com"
+)
+
+type service struct {
+	conn *websocket.Conn
+
+	marketByTicker map[string]ports.Market
+	feedChan       chan ports.PriceFeed
+	quitChan       chan struct{}
+}
+
+func NewKrakenPriceFeeder(args ...interface{}) (ports.PriceFeeder, error) {
+	if len(args) != 1 {
+		return nil, fmt.Errorf("invalid number of args")
+	}
+
+	markets, ok := args[0].([]ports.Market)
+	if !ok {
+		return nil, fmt.Errorf("unknown args type")
+	}
+
+	mktTickers := make([]string, 0, len(markets))
+	mktByTicker := make(map[string]ports.Market)
+	for _, mkt := range markets {
+		mktTickers = append(mktTickers, mkt.Ticker())
+		mktByTicker[mkt.Ticker()] = mkt
+	}
+
+	conn, err := connectAndSubscribe(mktTickers)
+	if err != nil {
+		return nil, err
+	}
+	conn.CloseHandler()
+
+	return &service{
+		conn:           conn,
+		marketByTicker: mktByTicker,
+		feedChan:       make(chan ports.PriceFeed),
+		quitChan:       make(chan struct{}, 1),
+	}, nil
+}
+
+func (k *service) Start() error {
+	defer func() {
+		close(k.feedChan)
+		close(k.quitChan)
+	}()
+
+	mustReconnect, err := k.start()
+	for mustReconnect {
+		log.WithError(err).Warn("connection dropped unexpectedly. Trying to reconnect...")
+
+		tickers := make([]string, 0, len(k.marketByTicker))
+		for ticker := range k.marketByTicker {
+			tickers = append(tickers, ticker)
+		}
+
+		var conn *websocket.Conn
+		conn, err = connectAndSubscribe(tickers)
+		if err != nil {
+			return err
+		}
+		k.conn = conn
+
+		log.Debug("connection and subscriptions re-established. Restarting...")
+		mustReconnect, err = k.start()
+	}
+
+	return err
+}
+
+func (k *service) Stop() {
+	k.quitChan <- struct{}{}
+}
+
+func (k *service) FeedChan() chan ports.PriceFeed {
+	return k.feedChan
+}
+
+func (k *service) start() (mustReconnect bool, err error) {
+	defer func() {
+		if rec := recover(); rec != nil {
+			mustReconnect = true
+		}
+	}()
+
+	for {
+		select {
+		case <-k.quitChan:
+			err = k.conn.Close()
+			return false, err
+		default:
+			_, message, err := k.conn.ReadMessage()
+			if err != nil {
+				if websocket.IsUnexpectedCloseError(err, websocket.CloseGoingAway, websocket.CloseAbnormalClosure) {
+					panic(err)
```

Comment:
not great to panic here deep in code, better to lift up eventually and handle in `cmd` maybe?

Reply:
is this the fix for #24 ? shoudnt we connect by ourself again via code?

Reply:
this is specifically intended. After debugging and testing how [this issue](https://support.kraken.com/hc/en-us/articles/360044504011-WebSocket-API-unexpected-disconnections-from-market-data-feeds) affects our service, basically what happens when Cloudflare closes the connection is that k.conn.ReadMessage() panics instead of returning the IsUnexpectedCloseError.
For this reason, we MUST recover the paniced code from the websocket pkg and signal. In case this kind of error is detected by this line, we panic like websocket would be to preserve the behavior.

Reply:
maybe better to add some commented line to explain why things are done this way, what do you think?

Reply:
yes then comment